network/peer/upgrader_test.go

```diff
@@ -0,0 +1,221 @@
+	"encoding/pem"
+	"math/big"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+
+	_ "embed"
+
+	"github.com/ava-labs/avalanchego/network/peer"
+	"github.com/ava-labs/avalanchego/staking"
+)
+
+// 8192RSA.pem is used here because it's too expensive
+// to generate an 8K bit RSA key under the time constraint of the weak Github CI runners.
+
+//go:embed 8192RSA.pem
+var fat8192BitRSAKey []byte
+
+func TestBlockClientsWithIncorrectRSAKeys(t *testing.T) {
+	privKey2048, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	privKey4096, err := rsa.GenerateKey(rand.Reader, 4096)
+	require.NoError(t, err)
+
+	block, _ := pem.Decode(fat8192BitRSAKey)
+	require.NotNil(t, block)
+
+	rsaFatKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	require.NoError(t, err)
+
+	privKey8192 := rsaFatKey.(*rsa.PrivateKey)
+
+	// Sanity check - ensure privKey8192 is indeed an 8192 RSA key
+	require.Equal(t, 8192, privKey8192.N.BitLen())
+
+	clientCert2048 := makeTLSCert(t, privKey2048)
+	clientCert4096 := makeTLSCert(t, privKey4096)
+	clientCert8192 := makeTLSCert(t, privKey8192)
+	clientCertBad := makeTLSCert(t, nonStandardRSAKey(t))
+
+	for _, testCase := range []struct {
+		description   string
+		clientTLSCert tls.Certificate
+		expectedErr   error
+	}{
+		{
+			description:   "Proper key size and private key - 2048",
+			clientTLSCert: clientCert2048,
+		},
+		{
+			description:   "Proper key size and private key - 4096",
+			clientTLSCert: clientCert4096,
+		},
+		{
+			description:   "Too big key",
+			clientTLSCert: clientCert8192,
+			expectedErr:   staking.ErrUnsupportedRSAModulusBitLen,
+		},
+		{
+			description:   "Improper public exponent",
+			clientTLSCert: clientCertBad,
+			expectedErr:   staking.ErrUnsupportedRSAPublicExponent,
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			serverKey, err := rsa.GenerateKey(rand.Reader, 2048)
+			require.NoError(t, err)
+
+			serverCert := makeTLSCert(t, serverKey)
+
+			config := peer.TLSConfig(serverCert, nil)
+
+			c := prometheus.NewCounter(prometheus.CounterOpts{})
+
+			// Initialize upgrader with a mock that fails when it's incremented.
+			failOnIncrementCounter := &mockPrometheusCounter{
+				Counter: c,
+				t:       t,
+				onIncrement: func() {
+					require.FailNow(t, "should not have invoked")
+				},
+			}
+			upgrader := peer.NewTLSServerUpgrader(config, failOnIncrementCounter)
+
+			clientConfig := tls.Config{
+				ClientAuth:         tls.RequireAnyClientCert,
+				InsecureSkipVerify: true, //#nosec G402
+				MinVersion:         tls.VersionTLS13,
+				Certificates:       []tls.Certificate{testCase.clientTLSCert},
+			}
+
+			listener, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+			defer listener.Close()
+
+			var wg sync.WaitGroup
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+				conn, err := listener.Accept()
+				require.NoError(t, err)
```

Comment:
I suspect that this would generate a race error in case an error is returned, since the go testing framework wouldn't allow you failing from an inner go-routine.
to "solve" this issue, I'd suggest using an error channel instead of the waitgroup and test it's result on the main go-routine.

Reply:
I don't think this causes a race condition, but the way that `t.FailNow()` works is by calling `panic` - which is normally caught by the testing harness managed when running a `Test` or in `t.Run(...)` but here that `panic` will escape the goroutine and cause the test to crash.

I doubt this is the only place in avalanchego where this could happen... so 🤷 on if we should fix it.